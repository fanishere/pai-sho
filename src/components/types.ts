export type Position = {
  x: number;
  y: number;
};

export type PieceData = {
  position: Position;
};

export interface PieceDataList {
  [index: string]: PieceData;
}

export type PieceGuideCandidates = {
  position: Position;
  offset: number;
};

export type PieceGuide = {
  position: Position;
  offset: number;
};

export type Side = 'red' | 'white';
