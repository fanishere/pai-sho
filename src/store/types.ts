export interface Position {
  x: number;
  y: number;
}

export interface PieceData {
  name: string;
  position: Position | null;
}

export type PieceGuide = {
  position: Position;
  offset: number;
};

export interface GameState {
  piecePositions: PieceData[];
  guidesVisible: PieceGuide[];
}

export const MOVE_PIECE = 'MOVE_PIECE';
export const SET_GUIDES = 'SET_GUIDES';
export const REMOVE_GUIDES = 'REMOVE_GUIDES';

export interface PieceMoveAction {
  type: typeof MOVE_PIECE;
  pieceName: string;
  position: Position;
}

export interface SetGuidesAction {
  type: typeof SET_GUIDES;
  guides: PieceGuide[];
}

export interface RemoveGuidesAction {
  type: typeof REMOVE_GUIDES;
}

export type GameActionTypes =
  | PieceMoveAction
  | SetGuidesAction
  | RemoveGuidesAction;
