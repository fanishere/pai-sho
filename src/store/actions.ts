import {
  MOVE_PIECE,
  PieceGuide,
  Position,
  REMOVE_GUIDES,
  SET_GUIDES
} from './types';
import { createAction } from '@reduxjs/toolkit';

export const movePiece = createAction(
  MOVE_PIECE,
  (pieceName: string, position: Position) => {
    return {
      payload: {
        type: MOVE_PIECE,
        pieceName: pieceName,
        position: position
      }
    };
  }
);

export const setGuides = createAction(SET_GUIDES, (guides: PieceGuide[]) => {
  return {
    payload: {
      type: SET_GUIDES,
      guides
    }
  };
});

export const removeGuides = createAction(REMOVE_GUIDES);
