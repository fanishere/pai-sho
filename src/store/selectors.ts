import { createSelector } from '@reduxjs/toolkit';
import { GameState, PieceData, PieceGuide } from './types';

export const piecesSelector = (state: GameState): PieceData[] => {
  return state.piecePositions;
};

export const pieceMapSelector = createSelector(
  piecesSelector,
  (pieces) => new Map(pieces.map((piece) => [piece.name, piece.position]))
);

export const guidesSelector = (state: GameState): PieceGuide[] => {
  return state.guidesVisible;
};
