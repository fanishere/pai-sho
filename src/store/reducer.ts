import { GameState } from './types';
import { createReducer } from '@reduxjs/toolkit';
import { movePiece, removeGuides, setGuides } from './actions';

const initialGameState: GameState = {
  piecePositions: [
    { name: 'rose', position: null },
    { name: 'second', position: null }
  ],
  guidesVisible: []
};

export const piecesReducer = createReducer(initialGameState, (builder) => {
  builder
    .addCase(movePiece, (state, action) => {
      return {
        ...state,
        piecePositions: state.piecePositions.map((piece) => {
          if (piece.name != action.payload.pieceName) {
            return piece;
          }
          return {
            ...piece,
            position: action.payload.position
          };
        })
      };
    })
    .addCase(setGuides, (state: GameState, action) => {
      return {
        ...state,
        guidesVisible: action.payload.guides
      };
    })
    .addCase(removeGuides, (state) => {
      return {
        ...state,
        guidesVisible: []
      };
    });
});
