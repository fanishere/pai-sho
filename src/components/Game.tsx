import React, { FunctionComponent, ReactElement } from 'react';
import { Circle, Group, Layer, Rect, Stage } from 'react-konva';
import { Piece } from './Piece/Piece';
import Konva from 'konva';
import Gate from './Gate';
import Garden from './Garden';
import PieceSelection from './PieceSelection';
import { useSelector, ReactReduxContext, Provider } from 'react-redux';
import { guidesSelector, piecesSelector } from '../store/selectors';
import rose from '../assets/rose.svg';

export type GameProps = {
  width: number;
  height: number;
  radius: number;
  gridCount: number;
};

const Game: FunctionComponent<GameProps> = (props: GameProps) => {
  const guidesVisible = useSelector(guidesSelector);
  // const isInsideGameBoard = (
  //   x1: number,
  //   x2: number,
  //   y1: number,
  //   y2: number
  // ): boolean => {
  //   return (
  //     Math.pow(x1 - x2, 2) + Math.pow(y1 - y2, 2) >= props.radius * props.radius
  //   );
  // };
  const clipFunc = (ctx: CanvasRenderingContext2D): void => {
    // clip a full circle of this.props.radius
    ctx.arc(0, 0, props.radius, 0, Math.PI * 2);
  };
  const layerRef:
    | React.RefObject<Konva.Layer>
    | null
    | undefined = React.createRef();
  const pieces = useSelector(piecesSelector);

  const smallGridWidth = props.radius / (props.gridCount / 2);
  const vertices: { x: number; y: number }[][] = Array(props.gridCount)
    .fill('')
    .map((_, i: number) => {
      return Array(props.gridCount)
        .fill('')
        .map((_, j) => {
          return {
            x: smallGridWidth * i - props.radius,
            y: smallGridWidth * j - props.radius
          };
        });
    });
  const boardPieces = pieces.map((piece) => {
    return (
      <Piece
        key={piece.name}
        name={piece.name}
        gridWidth={smallGridWidth}
        side={'red'}
        imgSrc={rose}
        spaces={3}
        position={piece.position}
      />
    );
  });

  return (
    <ReactReduxContext.Consumer>
      {({ store }): ReactElement => (
        <Stage width={window.innerWidth} height={window.innerHeight}>
          <Provider store={store}>
            <Layer
              offsetX={-props.radius}
              offsetY={-props.radius}
              ref={layerRef}
            >
              <Group clipFunc={clipFunc}>
                <Circle x={0} y={0} radius={props.radius} fill="pink"></Circle>

                {/*grids*/}
                {vertices.map((vertex, i) => {
                  return vertex.map((v, j) => {
                    return (
                      <Rect
                        // TODO make this key unique
                        key={i + j}
                        x={v.x}
                        y={v.y}
                        width={smallGridWidth}
                        height={smallGridWidth}
                        stroke="white"
                      />
                    );
                  });
                })}

                {/*guides*/}
                {guidesVisible.map((guide, i) => {
                  return (
                    <Circle
                      // TODO make this key unique
                      key={i}
                      x={guide.position.x}
                      y={guide.position.y}
                      stroke="white"
                      fill={'red'}
                      radius={5}
                    />
                  );
                })}

                <Gate
                  x={smallGridWidth * 2 - props.radius}
                  y={0}
                  gridWidth={smallGridWidth}
                  gameRadius={props.width}
                  rotation={90}
                ></Gate>
                <Gate
                  x={0}
                  y={props.radius - smallGridWidth * 2}
                  gridWidth={smallGridWidth}
                  gameRadius={props.width}
                  rotation={0}
                ></Gate>
                <Gate
                  x={0}
                  y={smallGridWidth * 2 - props.radius}
                  gridWidth={smallGridWidth}
                  gameRadius={props.width}
                  rotation={180}
                ></Gate>
                <Gate
                  x={props.radius - smallGridWidth * 2}
                  y={0}
                  gridWidth={smallGridWidth}
                  gameRadius={props.width}
                  rotation={270}
                ></Gate>
                <Garden
                  middle={[0, 0]}
                  vertex1={[smallGridWidth * 2 - props.radius, 0]}
                  vertex2={[0, smallGridWidth * 2 - props.radius]}
                  fill={'#fef3ed'}
                ></Garden>
                <Garden
                  middle={[0, 0]}
                  vertex1={[props.radius - smallGridWidth * 2, 0]}
                  vertex2={[0, smallGridWidth * 2 - props.radius]}
                  fill={'red'}
                ></Garden>
                <Garden
                  middle={[0, 0]}
                  vertex1={[props.radius - smallGridWidth * 2, 0]}
                  vertex2={[0, props.radius - smallGridWidth * 2]}
                  fill={'#fef3ed'}
                ></Garden>
                <Garden
                  middle={[0, 0]}
                  vertex1={[smallGridWidth * 2 - props.radius, 0]}
                  vertex2={[0, props.radius - smallGridWidth * 2]}
                  fill={'red'}
                ></Garden>
              </Group>
            </Layer>
            <PieceSelection pieces={boardPieces} />
          </Provider>
        </Stage>
      )}
    </ReactReduxContext.Consumer>
  );
};

export default Game;
