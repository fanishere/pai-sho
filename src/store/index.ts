import { piecesReducer } from './reducer';
import { configureStore } from '@reduxjs/toolkit';

export const store = configureStore({
  reducer: piecesReducer
});
