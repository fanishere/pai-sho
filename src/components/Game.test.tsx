import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import Game, { GameProps } from './Game';
import toJson from 'enzyme-to-json';
import 'jest-canvas-mock';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { GameState } from '../store/types';

describe('Mounted tests', () => {
  let wrapper: ReactWrapper;
  let game: ReactWrapper<GameProps>;
  const initialGameState: GameState = {
    piecePositions: [],
    guidesVisible: []
  };

  beforeEach(() => {
    const middlewares: any = [];
    const mockStore = configureStore(middlewares);
    const store = mockStore(initialGameState);

    wrapper = mount(
      <Provider store={store}>
        <Game height={300} width={300} radius={300} gridCount={18} />
      </Provider>
    );
    game = wrapper.find(Game);
  });

  describe('Test render', () => {
    it('renders Game with props', () => {
      expect(game).toHaveLength(1);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders with the props set', () => {
      expect(game.props().height).toEqual(300);
      expect(game.props().width).toEqual(300);
      expect(game.props().radius).toEqual(300);
      expect(game.props().gridCount).toEqual(18);
    });

    it('renders with 0 guides showing', () => {
      expect(game).toHaveLength(1);
    });
  });
});
