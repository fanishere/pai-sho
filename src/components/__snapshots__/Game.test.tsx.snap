// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted tests Test render renders Game with props 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Game
    gridCount={18}
    height={300}
    radius={300}
    width={300}
  >
    <ForwardRef
      height={768}
      width={1024}
    >
      <Stage
        forwardedRef={null}
        height={768}
        width={1024}
      >
        <div />
      </Stage>
    </ForwardRef>
  </Game>
</Provider>
`;
